aerogear/mobile-security-service-operator
# Install postgreSQL for the Mobile Security Service + refectories (AEROGEAR-8857)

## Motivation
https://issues.jboss.org/browse/AEROGEAR-8857

## What
* Create a new type/controller to install the database
* Add new commands to the makefile
* Implementing types for MobileSecurityServiceDB and  MobileSecurityServiceApp
* Update Readme 
* Refectories in the code in order to keep its maintenance ability and allow the possibility to extract the database to another operator/project that could be used as dependency here.  

## Why
* Allow users to get all installed and configured in their projects the Mobile Security Service REST+UI+Database with and without the Operator Lifecycle Manager (OLM) to manage its dependencies. 
* Allow the user install just the Mobile Security Service without database 
* Make easier we extract the DB logic for another operator to install just the PostgreSQL. 

## How
* Create a new CRD and Controler

```
$ operator-sdk add api --api-version=mobile-security-service.aerogear.com/v1alpha1 --kind=MobileSecurityServiceDB

$ operator-sdk add controller --api-version=mobile-security-service.aerogear.com/v1alpha1 --kind=MobileSecurityServiceDB
```

## Verification Steps

IMPORTANT: Update the version of the operator in the code and make build and make publish a new version in order to check this changes. (It is to ensure that you will use an tag which the changes performed in this PR)

1. Deploy the operator and check if it will work with success (command: `make deploy`)
2. Create a project and deploy the full app (command: `make deploy-app`)
3. Check that the app + database was installed and setup with success
4. Check that the env variables used by database and application came from the config map
5. Undeploy just the app (command: `make undeploy-app-only`), check if the app was removed and if the database still running with success
6. Remove the database deployment/pod by the OCP console and check if the operator after a few minutes were able to recreate another. ensure that the data used to create it is no longer the config map since it was removed by the command executed in the above test. 
7. Undeploy just the db  (command: `make undeploy-db-only`), check if the database and its objects will be removed from the project/namespace
8. Up/Install all with `make deploy-app` and check if all will be up with success, then create a route for the app project and check if the UI will appear. 
9. Run the commands `make undeploy-app` and `make undeploy`. Check if the app and database were removed from the project as the operator namespace removed from the cluster. 

## Checklist:


## Progress

- [] TODO

## Additional Notes



Following the verification performed/local tests. 

- [x] 1. Deploy the operator and check if it will work with success (command: `make deploy`)
<img width="1527" alt="Screenshot 2019-03-18 at 17 26 07" src="https://user-images.githubusercontent.com/7708031/54549946-09894a00-49a3-11e9-9da4-e07ad68677a5.png">
<img width="1465" alt="Screenshot 2019-03-18 at 17 26 33" src="https://user-images.githubusercontent.com/7708031/54549951-0beba400-49a3-11e9-99c4-88b4331e4997.png">

- [x] 2. Create a project and deploy the full app (command: `make deploy-app`)
<img width="1552" alt="Screenshot 2019-03-18 at 17 29 31" src="https://user-images.githubusercontent.com/7708031/54550177-6e44a480-49a3-11e9-9eef-b10645342d0b.png">

- [x] 3. Check that the app + database was installed and setup with success
<img width="1426" alt="Screenshot 2019-03-18 at 17 30 50" src="https://user-images.githubusercontent.com/7708031/54550263-992ef880-49a3-11e9-9728-f19ce604ab57.png">
<img width="1217" alt="Screenshot 2019-03-18 at 17 30 27" src="https://user-images.githubusercontent.com/7708031/54550264-992ef880-49a3-11e9-8512-76d73bdfe0e0.png">

- [x] 4. Check that the env variables used by database and application came from the config map
<img width="1530" alt="Screenshot 2019-03-18 at 17 31 39" src="https://user-images.githubusercontent.com/7708031/54550329-b5cb3080-49a3-11e9-9b3d-011ada1c5c99.png">
<img width="1475" alt="Screenshot 2019-03-18 at 17 31 28" src="https://user-images.githubusercontent.com/7708031/54550330-b5cb3080-49a3-11e9-91f9-325075f9b5d2.png">

- [x] 5. Undeploy just the app (command: `make undeploy-app-only`), check if the app was removed and if the database still running with success

<img width="1541" alt="Screenshot 2019-03-18 at 17 32 11" src="https://user-images.githubusercontent.com/7708031/54550370-c8de0080-49a3-11e9-9bdc-2a894ed5f5cc.png">

- [x] 6. Remove the database deployment/pod by the OCP console and check if the operator after a few minutes were able to recreate another. ensure that the data used to create it is no longer the config map since it was removed by the command executed in the above test. 

<img width="1538" alt="Screenshot 2019-03-18 at 17 38 36" src="https://user-images.githubusercontent.com/7708031/54550830-b1534780-49a4-11e9-9bcc-d3a096b77b47.png">
<img width="1534" alt="Screenshot 2019-03-18 at 17 38 29" src="https://user-images.githubusercontent.com/7708031/54550831-b1534780-49a4-11e9-9b5a-08f6fd269170.png">

- [x] 7. Undeploy just the DB  (command: `make undeploy-db-only`), check if the database and its objects will be removed from the project/namespace

```
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/aerogear/mobile-security-service-operator (DB) $ make undeploy-db-only
Undeploying Mobile Security Service and Database from the project:
oc delete -f deploy/crds/mobile-security-service_v1alpha1_mobilesecurityservicedb_cr.yaml
mobilesecurityservicedb.mobile-security-service.aerogear.com "mobile-security-service-db" deleted
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/aerogear/mobile-security-service-operator (DB) $ get oc all
-bash: get: command not found
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/aerogear/mobile-security-service-operator (DB) $ 
```

- [x] 8. Up/Install all with `make deploy-app` and check if all will be up with success, then create a route for the app project and check if the UI will appear. 

<img width="1536" alt="Screenshot 2019-03-18 at 17 44 08" src="https://user-images.githubusercontent.com/7708031/54551202-84536480-49a5-11e9-8a8d-4c425b62cd5c.png">
<img width="1097" alt="Screenshot 2019-03-18 at 17 44 02" src="https://user-images.githubusercontent.com/7708031/54551203-84536480-49a5-11e9-9405-01d5ad50b859.png">

- [x] 9. Run the commands `make undeploy-app` and `make undeploy`. Check if the app and database were removed from the project as the operator namespace removed from the cluster.
